src/plugins/utils/chart-render.ts

```diff
@@ -0,0 +1,145 @@
+    type: 'smooth',
+};
+
+export const renderChart = ({
+    engine,
+    points,
+    style,
+    min,
+    max,
+}: {
+    engine: RenderEngine | OffscreenRenderEngine;
+    points: ChartPoints;
+    min: number;
+    max: number;
+    style?: Partial<ChartStyle>;
+}) => {
+    const resolvedStyle = {
+        ...defaultChartStyle,
+        ...(style ?? {}),
+    };
+
+    engine.setCtxValue('strokeStyle', resolvedStyle.lineColor);
+    engine.setCtxValue('fillStyle', resolvedStyle.fillColor);
+    engine.setCtxValue('lineWidth', resolvedStyle.lineWidth);
+    engine.callCtx('setLineDash', resolvedStyle.lineDash);
+
+    engine.ctx.beginPath();
+
+    const levelHeight = (engine.height - engine.charHeight - 4) / (max - min);
+
+    if (points.length > 1) {
+        const xy = points.map(([time, level]) => [
+            engine.timeToPosition(time),
+            castLevelToHeight(level, min, levelHeight, engine.height),
+        ]);
+
+        engine.ctx.moveTo(xy[0][0], engine.height);
+        engine.ctx.lineTo(xy[0][0], xy[0][1]);
+
+        if (resolvedStyle.type === 'smooth' || !resolvedStyle.type) {
+            for (let i = 1; i < xy.length - 2; i++) {
+                const xc = (xy[i][0] + xy[i + 1][0]) / 2;
+                const yc = (xy[i][1] + xy[i + 1][1]) / 2;
+
+                engine.ctx.quadraticCurveTo(xy[i][0], xy[i][1], xc, yc);
+            }
+
+            const preLast = xy[xy.length - 2];
+            const last = xy[xy.length - 1];
+
+            engine.ctx.quadraticCurveTo(preLast[0], preLast[1], last[0], last[1]);
+            engine.ctx.quadraticCurveTo(last[0], last[1], last[0], engine.height);
+        } else if (resolvedStyle.type === 'line') {
+            for (let i = 1; i < xy.length; i++) {
+                engine.ctx.lineTo(xy[i][0], xy[i][1]);
+            }
+        } else if (resolvedStyle.type === 'bar') {
+            for (let i = 0; i < xy.length; i++) {
+                const currentPoint = xy[i];
+                const prevPoint = xy[i - 1] || currentPoint;
+                const nextPoint = xy[i + 1];
+
+                const barWidthLeft = (currentPoint[0] - prevPoint[0]) / 2;
+                const barWidthRight = nextPoint ? (nextPoint[0] - currentPoint[0]) / 2 : barWidthLeft;
+
+                engine.ctx.lineTo(prevPoint[0] + barWidthLeft, currentPoint[1]);
+                engine.ctx.lineTo(currentPoint[0] + barWidthRight, currentPoint[1]);
+
+                if (nextPoint) {
+                    engine.ctx.lineTo(currentPoint[0] + barWidthRight, nextPoint[1]);
+                } else {
+                    engine.ctx.lineTo(currentPoint[0] + barWidthRight, engine.height);
+                }
+            }
+
+            engine.ctx.lineTo(xy[xy.length - 1][0], engine.height);
+        }
+    }
+
+    engine.ctx.closePath();
+
+    engine.ctx.stroke();
+    engine.ctx.fill();
+};
+
+export const binarySearch = (array: ChartPoints, value: number, outside: boolean = true) => {
```

Comment:
its called binarySearch, but it is very coupled with chartpoints array structure
name is generic but implementation is not
maybe we can then name it more appropriate?

also, what do you think about the rule to enforce declaring return types on functions?

Reply:
In most cases, the typescript does a good job of inferring the type of the result of a function. Therefore, such a rule would be redundant, in my opinion.)

Reply:
> In most cases, the typescript does a good job of inferring the type of the result of a function. Therefore, such a rule would be redundant, in my opinion.)

honestly I just suggested it because I got burned few times because of that, so we in my team decided to enforce it by the rule
also a short video about it https://www.youtube.com/watch?v=Gcr4t6cH-lU
ofcourse, I might just be opinionated here

Reply:
Well, in general, I agree of course. But for example, I always look at function annotations in my ide and there the typescript always displays the correct type, but if it type is large and complex and not described before, then I completely agree that it should be taken out and described. But for example, if it is a simple boolean type or some specific complex type, which is already infered by some variable inside the function body, then I think it's not worth it. Perhaps you can set up the rules so that it would only throw an error if the function returns a complex object type that has not yet been declared?